apollo-router/src/plugins/telemetry/metrics/mod.rs

```diff
@@ -487,52 +480,95 @@ impl AttributesForwardConf {
+    pub(crate) apollo_meter_provider_builder: opentelemetry::sdk::metrics::MeterProviderBuilder,
+    pub(crate) prometheus_meter_provider: Option<opentelemetry::sdk::metrics::MeterProvider>,
+    pub(crate) custom_endpoints: MultiMap<ListenAddr, Endpoint>,
+    pub(crate) apollo_metrics_sender: Sender,
+    pub(crate) resource: Resource,
 }
 
-impl MetricsBuilder {
-    pub(crate) fn exporters(&mut self) -> Vec<MetricsExporterHandle> {
-        std::mem::take(&mut self.exporters)
-    }
-    pub(crate) fn meter_provider(&mut self) -> AggregateMeterProvider {
-        AggregateMeterProvider::new(std::mem::take(&mut self.meter_providers))
-    }
-    pub(crate) fn custom_endpoints(&mut self) -> MultiMap<ListenAddr, Endpoint> {
-        std::mem::take(&mut self.custom_endpoints)
-    }
-
-    pub(crate) fn apollo_metrics_provider(&mut self) -> Sender {
-        self.apollo_metrics.clone()
+struct ConfigResourceDetector(MetricsCommon);
+
+impl ResourceDetector for ConfigResourceDetector {
+    fn detect(&self, _timeout: Duration) -> Resource {
+        let mut resource = Resource::new(
+            vec![
+                self.0.service_name.clone().map(|service_name| {
+                    KeyValue::new(
+                        opentelemetry_semantic_conventions::resource::SERVICE_NAME,
+                        service_name,
+                    )
+                }),
+                self.0.service_namespace.clone().map(|service_namespace| {
+                    KeyValue::new(
+                        opentelemetry_semantic_conventions::resource::SERVICE_NAMESPACE,
+                        service_namespace,
+                    )
+                }),
+            ]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<_>>(),
+        );
+        resource = resource.merge(&mut Resource::new(
+            self.0
+                .resources
+                .clone()
+                .into_iter()
+                .map(|(k, v)| KeyValue::new(k, v)),
+        ));
+        resource
     }
 }
 
 impl MetricsBuilder {
-    fn with_exporter<T: Send + Sync + 'static>(mut self, handle: T) -> Self {
-        self.exporters.push(Box::new(handle));
-        self
-    }
-
-    fn with_meter_provider<T: MeterProvider + Send + Sync + 'static>(
-        mut self,
-        meter_provider: T,
-    ) -> Self {
-        self.meter_providers.push(Arc::new(meter_provider));
-        self
-    }
-
-    fn with_custom_endpoint(mut self, listen_addr: ListenAddr, endpoint: Endpoint) -> Self {
-        self.custom_endpoints.insert(listen_addr, endpoint);
-        self
-    }
+    pub(crate) fn new(config: &Conf) -> Self {
+        let metrics_common_config = config
+            .metrics
+            .clone()
+            .and_then(|m| m.common)
+            .unwrap_or_default();
+
+        let mut resource = Resource::from_detectors(
+            Duration::from_secs(0),
```

Comment:
Not sure about this. Setting this to a positive amount of time implies that it may block for that amount of time, and we'll have to move execution to a blocking task.

Resource::default() also uses duration of zero under the hood.